pkg/actuators/machine/machine_scope_test.go

```diff
@@ -286,3 +290,55 @@ func TestPatchMachine(t *testing.T) {
 		})
 	}
 }
+
+func TestGetCustomDomainFromDHCP(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	mockAWSClient := mockaws.NewMockClient(mockCtrl)
+	dhcpID := "someID"
+	expectedDomains := "openshift.com openshift.io"
+	mockAWSClient.EXPECT().DescribeVpcs(gomock.Any()).Return(&ec2.DescribeVpcsOutput{
+		Vpcs: []*ec2.Vpc{
+			&ec2.Vpc{DhcpOptionsId: &dhcpID},
+		},
+	}, nil).AnyTimes()
+
+	mockAWSClient.EXPECT().DescribeDHCPOptions(gomock.Any()).Return(&ec2.DescribeDhcpOptionsOutput{
+		DhcpOptions: []*ec2.DhcpOptions{
+			&ec2.DhcpOptions{
+				DhcpConfigurations: []*ec2.DhcpConfiguration{
+					&ec2.DhcpConfiguration{
+						Key: &dhcpDomainKeyName,
+						Values: []*ec2.AttributeValue{
+							&ec2.AttributeValue{
+								Value: &expectedDomains,
+							},
+						},
+					},
+				},
+			},
+		},
+	}, nil).AnyTimes()
+
+	testCases := []struct {
+		description string
+		expected    []string
+	}{
+		{
+			description: "Returns two",
+			expected:    []string{"openshift.com", "openshift.io"},
+		},
+	}
+
```

Comment:
it seems like there are several conditions where `getCustomDomainFromDHCP` can return `nil, nil`, does it make sense to add those as unit tests as well?

Reply:
Not really, IMO.

Reply:
can you elaborate why not? as far as I can see it definitely make sense and has value to cover more code paths, e.g description: "Returns one", description: "Returns empty"?

Reply:
I'm not a huge fan of mocking up fake cloud responses that don't materially drive decision making in our code.  If the cloud sends back some empty stuff, cool, we don't have any domains.  This case (returning nil,nil) is actually covered by several other tests because we don't mock any return from DescribeVPCS or DescribeDhcpOptions.

The only case where I can see something actually being empty is vpc.Vpcs[0].DhcpOptionsId.  It's maybe possible to have a VPC without a DHCP option set assigned (and this is the case that is covered via other unit tests).  It's also probably possible to have a DHCP option set assigned without any actual options, and possibly one defined with some options but nothing for domain-name, but does that actually happen ever?  Probably not, or at least not in a VPC you expect to work.  And even in that very limited case, the absence doesn't affect us.  We're not going to fail on missing info.

Since we're only driving this function with a single input (vpcID) which is entirely contrived in testing, adding a bunch of statements to simulate the AWS API seems pointless.  Firstly, there's no guarantee we're going to mock all the possible responses, let alone mock them correctly.  Secondly, we're validating the output of this function as a valid input to another function (extractNodeAddresses) via other tests, albeit the nil,nil case.  We're validating that this function will return a list of strings.  We validate in extractNodeAddresses that a list of strings does what we want.

The more important tests would be validating the error conditions, because that will actually break existing users.  But again, that's just mocking an established library and API, so there's no value in doing that.  If this code breaks, it will 'fail open' unless we segfault, which we should be defended against properly now.

Reply:
that makes sense to me. my main concern with the original question was just to make sure we are exercising all the possible return paths in that function. i wouldn't expect a full mock of the upstream api, just enough to cover the various return paths. but i get what you are saying about the varied paths that could lead to those error conditions based on the range of possible inputs.

Reply:
>If the cloud sends back some empty stuff, cool, we don't have any domains...

This func has 5 return calls and we only certainly know it does what it's meant for one of single return path and only in the case we expect "Returns two".

We don't know what the func does if DhcpOptions legitimately has only 1 domain or has no domains. And so we don't know if any future change to the implementation of the func would break that behaviour.

To me seems a lot of value we could get for the price of adding a few lines including 2 more testCases.

Any ways not a blocker to me...